go/vt/vtgate/planbuilder/physical/route_planning.go

```diff
@@ -238,9 +238,87 @@ func createRoute(ctx *plancontext.PlanningContext, table *abstract.QueryTable, s
 		}
 	}
 
+	if plan.RouteOpCode == engine.Scatter && len(table.Predicates) > 0 {
+		// If we have a scatter query, it's worth spending a little extra time seeing if we can't improve it
+		for _, pred := range table.Predicates {
+			rewritten := tryRewriteOrToIn(pred)
+			if rewritten != nil {
+				err = plan.UpdateRoutingLogic(ctx, rewritten)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
 	return plan, nil
 }
 
+func tryRewriteOrToIn(expr sqlparser.Expr) sqlparser.Expr {
+	rewrote := false
+	newPred := sqlparser.Rewrite(sqlparser.CloneExpr(expr), func(cursor *sqlparser.Cursor) bool {
+		_, ok := cursor.Node().(*sqlparser.OrExpr)
+		return ok
+	}, func(cursor *sqlparser.Cursor) bool {
+		// we are looking for the pattern WHERE c = 1 or c = 2
+		switch or := cursor.Node().(type) {
+		case *sqlparser.OrExpr:
+			lftCmp, ok := or.Left.(*sqlparser.ComparisonExpr)
+			if !ok {
+				return true
+			}
+			rgtCmp, ok := or.Right.(*sqlparser.ComparisonExpr)
+			if !ok {
+				return true
+			}
+
+			col, ok := lftCmp.Left.(*sqlparser.ColName)
+			if !ok || !sqlparser.EqualsExpr(lftCmp.Left, rgtCmp.Left) {
+				return true
+			}
```

Comment:
Can we do this for any type of node? For instance for `c+1`. If they are equal on the left and right comparisons, it should be safe to convert them to an `IN`

Reply:
technically and semantically we could, but since these predicates are only used to update the planning logic, and if the LHS is not a bare column, we can't use it for a vindex (it's not [sargable](https://en.wikipedia.org/wiki/Sargable))